### PR TITLE
feat: Migrate to React Native 0.71

### PR DESCRIPTION
**Pull request by mrousavy**
* Uses React Native core as prefabs on Android (binary compatibility)
* Uses new buildscript (build.gradle and CMakeLists are now significantly smaller, no more downloading, unzipping, header importing or cpp compiling - we now use prefabs for JSI)
* Supports running on the new architecture (Fabric, TurboModules)

> ⚠️ Note: This is a breaking change. react-native-mmkv now only builds on RN 0.71 and above using the new architecture. For the sake of simplicity, it is **not backwards compatible**.

### CODE REVIEW

File: example/package.json [hunk 1]
```diff
@@ -15,26 +15,27 @@
...
-    "@types/react-native": "^0.64.13",
-    "@typescript-eslint/eslint-plugin": "^4.33.0",
+    "@babel/core": "^7.20.2",
+    "@babel/runtime": "^7.20.1",
+    "@react-native-community/eslint-config": "^3.2.0",
+    "@react-native-community/eslint-plugin": "^1.3.0",
+    "@tsconfig/react-native": "^2.0.2",
+    "@types/jest": "^29.2.2",
+    "@types/react": "^18.0.25",
+    "@types/react-native": "^0.70.6",
```
Comment (NickGerleman) [inline example/package.json:30]: What happens if you remove this line?

Comment (mrousavy) [reply example/package.json:30]: @NickGerleman you mean all 4 packages, or just the @types/react-native package?

Comment (kelset) [reply example/package.json:30]: I think he means just the react-native ones; in 0.71 we bundle them into the node_module

Comment (NickGerleman) [reply example/package.json:30]: Yep. Just @types/react-native which should also now be builtin with newer types.

Comment (mrousavy) [reply example/package.json]: Just removed 'em - I think it works! :)

**Comment by meypod** [comment]
any ETA on this ? It's really appreciated
I'm kinda in the middle of migration to 0.71 due to some changes and now I see `react-native-mmkv` cannot compile

**Comment by meypod** [comment]
Oh, nice! happy to hear it's ready

**Comment by focux** [comment]
Is there any plan to support Paper for those who still haven't migrated to the new architecture?

**Comment by retyui** [comment]
React native 0.71 + new arch failed:

```
BUILD FAILED in 3m 38s

error Failed to install the app. Make sure you have the Android development environment set up: https://reactnative.dev/docs/environment-setup.
Error: Command failed: ./gradlew app:installDebug -PreactNativeDevServerPort=8081
[CXX1101] NDK at /home/i/Android/Sdk/ndk/23.1.7779620 did not have a source.properties file

FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':react-native-mmkv:configureCMakeDebug[arm64-v8a]'.
> [CXX1405] error when building with cmake using /home/i/tmp/RN71/node_modules/react-native-mmkv/android/CMakeLists.txt: -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /home/i/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/clang - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /home/i/Android/Sdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Configuring incomplete, errors occurred!
  See also "/home/i/tmp/RN71/node_modules/react-native-mmkv/android/.cxx/Debug/t38e92d3/arm64-v8a/CMakeFiles/CMakeOutput.log".
  
  C++ build system [configure] failed while executing:
      /home/i/Android/Sdk/cmake/3.18.1/bin/cmake \
        -H/home/i/tmp/RN71/node_modules/react-native-mmkv/android \
        -DCMAKE_SYSTEM_NAME=Android \
        -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
        -DCMAKE_SYSTEM_VERSION=21 \
        -DANDROID_PLATFORM=android-21 \
        -DANDROID_ABI=arm64-v8a \
        -DCMAKE_ANDROID_ARCH_ABI=arm64-v8a \
        -DANDROID_NDK=/home/i/Android/Sdk/ndk/23.1.7779620 \
        -DCMAKE_ANDROID_NDK=/home/i/Android/Sdk/ndk/23.1.7779620 \
        -DCMAKE_TOOLCHAIN_FILE=/home/i/Android/Sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake \
        -DCMAKE_MAKE_PROGRAM=/home/i/Android/Sdk/cmake/3.18.1/bin/ninja \
        "-DCMAKE_CXX_FLAGS=-O2 -frtti -fexceptions -Wall -Wno-unused-variable -fstack-protector-all" \
        -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/home/i/tmp/RN71/node_modules/react-native-mmkv/android/build/intermediates/cxx/Debug/t38e92d3/obj/arm64-v8a \
        -DCMAKE_RUNTIME_OUTPUT_DIRECTORY=/home/i/tmp/RN71/node_modules/react-native-mmkv/android/build/intermediates/cxx/Debug/t38e92d3/obj/arm64-v8a \
        -DCMAKE_BUILD_TYPE=Debug \
        -DCMAKE_FIND_ROOT_PATH=/home/i/tmp/RN71/node_modules/react-native-mmkv/android/.cxx/Debug/t38e92d3/prefab/arm64-v8a/prefab \
        -B/home/i/tmp/RN71/node_modules/react-native-mmkv/android/.cxx/Debug/t38e92d3/arm64-v8a \
        -GNinja \
        -DAPP_BUILD_DIR=/home/i/tmp/RN71/android/app/build \
        -DREACT_NATIVE_DIR=/home/i/tmp/RN71/node_modules/react-native \
        -DANDROID_STL=c++_shared
    from /home/i/tmp/RN71/node_modules/react-native-mmkv/android
  CMake Warning at /home/i/Android/Sdk/ndk/23.1.7779620/build/cmake/android-legacy.toolchain.cmake:416 (message):
    An old version of CMake is being used that cannot automatically detect
    compiler attributes.  Compiler identification is being bypassed.  Some
    values may be wrong or missing.  Update to CMake 3.19 or newer to use
    CMake's built-in compiler identification.
  Call Stack (most recent call first):
    /home/i/Android/Sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake:55 (include)
    /home/i/Android/Sdk/cmake/3.18.1/share/cmake-3.18/Modules/CMakeDetermineSystem.cmake:93 (include)
    CMakeLists.txt:1 (project)
  
  
  CMake Warning at /home/i/Android/Sdk/ndk/23.1.7779620/build/cmake/android-legacy.toolchain.cmake:416 (message):
    An old version of CMake is being used that cannot automatically detect
    compiler attributes.  Compiler identification is being bypassed.  Some
    values may be wrong or missing.  Update to CMake 3.19 or newer to use
    CMake's built-in compiler identification.
  Call Stack (most recent call first):
    /home/i/Android/Sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake:55 (include)
    /home/i/tmp/RN71/node_modules/react-native-mmkv/android/.cxx/Debug/t38e92d3/arm64-v8a/CMakeFiles/3.18.1-g262b901/CMakeSystem.cmake:6 (include)
    /home/i/tmp/RN71/node_modules/react-native-mmkv/android/.cxx/Debug/t38e92d3/arm64-v8a/CMakeFiles/CMakeTmp/CMakeLists.txt:2 (project)
  
  
  CMake Warning at /home/i/Android/Sdk/ndk/23.1.7779620/build/cmake/android-legacy.toolchain.cmake:416 (message):
    An old version of CMake is being used that cannot automatically detect
    compiler attributes.  Compiler identification is being bypassed.  Some
    values may be wrong or missing.  Update to CMake 3.19 or newer to use
    CMake's built-in compiler identification.
  Call Stack (most recent call first):
    /home/i/Android/Sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake:55 (include)
    /home/i/tmp/RN71/node_modules/react-native-mmkv/android/.cxx/Debug/t38e92d3/arm64-v8a/CMakeFiles/3.18.1-g262b901/CMakeSystem.cmake:6 (include)
    /home/i/tmp/RN71/node_modules/react-native-mmkv/android/.cxx/Debug/t38e92d3/arm64-v8a/CMakeFiles/CMakeTmp/CMakeLists.txt:2 (project)
  
  
  CMake Error at CMakeLists.txt:16 (add_subdirectory):
    add_subdirectory given source "../MMKV/Core" which is not an existing
    directory.


```

**Comment by Noitham** [comment]
I got this error when building android using this mmkv version (react-native 0.71) - new arch enabled:

As reference:

```
"@react-native-community/slider": "4.4.0",
"react-native-mmkv": "mrousavy/react-native-mmkv#feat/migrate-rn-71",
```


```
> Task :app:mergeLibDexDebug FAILED

...
...

com.android.builder.dexing.DexArchiveMergerException: Error while merging dex archives: 
Learn how to resolve the issue at https://developer.android.com/studio/build/dependencies#duplicate_classes.

Type com.facebook.react.viewmanagers.RNCSliderManagerDelegate is defined multiple times: /Users/.../Code/.../...node_modules/@react-native-community/slider/android/build/.transforms/fecc51725eb5ee62117cebd51726ea5a/transformed/classes/classes.dex, /Users/Users/.../Code/Users/.../Users/.../node_modules/react-native-mmkv/android/build/.transforms/5c9ea4485730d687eda318fd506e13ee/transformed/classes/classes.dex
        at com.android.builder.dexing.D8DexArchiveMerger.getExceptionToRethrow(D8DexArchiveMerger.java:151)
        at com.android.builder.dexing.D8DexArchiveMerger.mergeDexArchives(D8DexArchiveMerger.java:138)

...
...

Caused by: com.android.tools.r8.CompilationFailedException: Compilation failed to complete, origin: /Users/.../Code/.../.../node_modules/@react-native-community/slider/android/build/.transforms/fecc51725eb5ee62117cebd51726ea5a/transformed/classes/classes.dex

```

**Pull request merged**